### PR TITLE
issue-479/collapse-view-header

### PR DESCRIPTION
**Pull request by lowlandjuju**
Resoves #479

### CODE REVIEW

**Comment by richardolsson** [comment]
Awesome work!

I'll let @djbusstop review the code because I think she'll have more valuable feedback. But as far as the design goes, here are some comments (based on my understanding of the outcome from the discussion that we had in #479):

* I think the collapse/expand toggle should go in the top right, next to where the search button will eventually land (#454)
* I think the title could go away entirely since it's already in the breadcrumbs
* Maybe the title in the breadcrumbs should be made bold or something, when collapsed?

See screenshot below (where I've pointed out what I'm talking about):

Screenshot 2021-11-29 at 11 18 31

Also, the change you've made to `TabbedLayout` affects other parts of the system. I think that's fine, this is actually a nice general feature IMO. But in other parts of the system (and maybe in _Views_ as well, in the future) there can be subtitles and action buttons. I think they should be hidden in the collapsed state.

In the screenshot below "June 12 - August 23, 2021" is the subtitle, and the "public page" text and cogwheels are action buttons (the latter will be replaced with a vertical ellipsis eventually, per the wireframes).

image

**Comment by lowlandjuju** [comment]
@richardolsson all good points - I have moved the location of the collapse button, made it hide the title, and all other header components except for the breadcrumbs.

**Comment by lowlandjuju** [comment]
@richardolsson @djbusstop The Playwright tests pass fine on my local environment, so I'm going to leave this for you to check out in your own time 👍 - maybe you'll see something obvious in the CI trace x

**Comment by lowlandjuju** [comment]
Ignore that last comment - playwright passed on attempt #4.

File: src/types/index.ts
Comment (djbusstop) [inline src/types/index.ts:24]: I like that this lives here now!

File: src/components/BreadcrumbTrail.tsx
Comment (djbusstop) [inline src/components/BreadcrumbTrail.tsx:46]: Oooooooooo :+1:

File: src/fetching/getBreadcrumbs.ts
Comment (djbusstop) [inline src/fetching/getBreadcrumbs.ts:8]: In this handler, we should throw an error if there's a non 200 status. Then, in the file `BreadcrumbTrail.tsx` we could conditionally show the trail only if react query has `query.isSuccess === true`. The way it works now works but it isn't so clear about errors, so this is just a code improvement. Not required. I think it's great you worked on improving this code anyway!

File: src/components/layout/organize/TabbedLayout.spec.tsx
Comment (djbusstop) [inline src/components/layout/organize/TabbedLayout.spec.tsx:4]: Maybe we should expose the feature you implemented as a prop which can be passed to the layout component? I wonder if we want it by default on every page? @richardolsson what do you think?

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/TabbedLayout.tsx:95]: This should be localised! You can do this by adding an english localised text in a `.yml` file in `src/locale/` you can put it wherever it makes sense. Maybe `misc/en.yml`. You can then implement this with `<FormattedMessage id={...}/>`

Then, for testing, our test runner doesn't localise text but shows the id used to look up text. So you could assert that the id of the text you want is rendered.

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/TabbedLayout.tsx:31]: I wonder why that's needed

File: src/components/BreadcrumbTrail.tsx
Comment (djbusstop) [inline src/components/BreadcrumbTrail.tsx:39]: Really nice feature :)

**Review by djbusstop — CHANGES_REQUESTED** [review]
The only really necessary thing I mention is the localisation. Every other comment is just an idea to consider. 

Using the feature is really really nice! It looks great :+1:

File: src/fetching/getBreadcrumbs.ts
Comment (lowlandjuju) [reply src/fetching/getBreadcrumbs.ts:8]: I agree it looks like there should be error handling - it looks like the `/fetching` folder in general could do with attention in this respect! Maybe all fetch-related error handling could be incorporated into a separate issue? :-)

File: src/components/layout/organize/TabbedLayout.spec.tsx
Comment (lowlandjuju) [reply src/components/layout/organize/TabbedLayout.spec.tsx:4]: Yes I was thinking about a 'collapsable' prop... Will leave for now

Somewhat related musing: had a thought about a container component that can allow any child component to be expanded to fullscreen, something like <Fullscreenable><Big.Table.Component /></Fullscreenable>, allowing any view a user might conceivably want to make big, to be made big...
I.e. the focus would be on expanding the view to cover other things, rather than moving the other things out of the way.

File: src/components/layout/organize/TabbedLayout.tsx
Comment (lowlandjuju) [reply src/components/layout/organize/TabbedLayout.tsx:31]: It wasn't 😅

File: src/components/layout/organize/TabbedLayout.tsx
Comment (lowlandjuju) [reply src/components/layout/organize/TabbedLayout.tsx:95]: Can't believe I forgot about this, since I had actually spent some time reflecting on how great it is that this project is already internationalised (at least in 2 languages at this point anyway...)
Fixed in latest commit :)

File: src/fetching/getBreadcrumbs.ts
Comment (richardolsson) [reply src/fetching/getBreadcrumbs.ts:8]: Yeah, changing all "fetching" modules for consistency is on our radar along with a bigger refactor of queries, but we also try to make sure any new ones we create are correctly compatible with `react-query` (e.g. throw an error if they fail). So it would be great if you could make this one compatible as well, i.e. add these three lines before parsing the response:

```ts
if (!bRes.ok) {
  throw new APIError('GET', url);
}
```

Comment (lowlandjuju) [reply src/fetching/getBreadcrumbs.ts:8]: Will do!

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [reply src/components/layout/organize/TabbedLayout.tsx:36]: @lowlandjuju Does that mean we should take it out?

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/TabbedLayout.tsx:102]: I think this is a case for a bit more complex use of the format strings in our internationalisation library.

First off I think you might want to use the `<FormattedMessage />` component. 

Then the text string can handle conditions IN the string, by accessing an option passed to the component

```
<FormattedMessage
  id="layout.organize.header.collapseButton"
  options={{
    collapsed
  }}
/>
```

Locale string:
```
collapseButton:  {collapsed, select, true {expand} false {collapse}} header
```



This is using the "ICU Message Format", I found an example of the switch select on this page: https://lokalise.com/blog/complete-guide-to-icu-message-format/#Switch_with_select. The contrived example is a stupid gender / pronouns one :D At least it has "they"!

Comment (djbusstop) [reply src/components/layout/organize/TabbedLayout.tsx:102]: This may not be a perfect example but this is kinda the direction to go.

Comment (richardolsson) [reply src/components/layout/organize/TabbedLayout.tsx:102]: Yeah, concatenating strings like that will not work in some languages. I imagine German probably has the verb (collapse/expand) after the noun (header), for example.

Either do what @djbusstop suggests and use ICU `select`, or create two distinct messages that contain the entire string, and switch between the two in TS (i.e what @lowlandjuju already did, except the concatenation). This might actually be easier, and result both in less code and message text that is easier for translators to translate:

```yml
header:
    collapseButton:
        collapse: collapse header
        expand: expand header
```

Comment (lowlandjuju) [reply src/components/layout/organize/TabbedLayout.tsx:102]: Prefer the latter solution - repeating the word header is less verbose overall

File: src/components/layout/organize/TabbedLayout.tsx
Comment (lowlandjuju) [reply src/components/layout/organize/TabbedLayout.tsx:36]: yes - I thought I had but got distracted

File: src/components/layout/organize/TabbedLayout.tsx
Comment (djbusstop) [inline src/components/layout/organize/TabbedLayout.tsx:100]: Ok I am sorry to be super pedantic but for one there is the component `<FormattedMessage />` which handles this which we tend to use when putting localised text in JSX.

Comment (lowlandjuju) [reply src/components/layout/organize/TabbedLayout.tsx]: 2 instances of useIntl hook replaced with ´<FormattedMessage />´  😉

**Review by djbusstop — DISMISSED** [review]

**Comment by djbusstop** [comment]
I added one small commit @lowlandjuju! This is ready to merge once tests pass

**Review by richardolsson — APPROVED** [review]
The added commit prompts another review to unblock merging. All looks good so here's approval. 😊 

We usually let the (primary) author of a PR do the merge, so that's you @lowlandjuju. Feel free to merge at will. 🎉

**Pull request merged**
